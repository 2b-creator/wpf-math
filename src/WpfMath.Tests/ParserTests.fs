module WpfMath.Tests.ParserTests

open Xunit

open WpfMath.Exceptions
open WpfMath.Tests.ApprovalTestUtils
open WpfMath.Tests.Utils

[<Fact>]
let ``2+2 should be parsed properly`` () =
    verifyParseResult "2+2"

[<Theory>]
[<InlineData("(", ")", "(", ")")>]
[<InlineData("[", "]", "lbrack", "rbrack")>]
[<InlineData("{", "}", "lbrace", "rbrace")>]
[<InlineData("<", ">", "langle", "rangle")>]
let delimiters (left : string, right : string, lResult : string, rResult : string) =
    verifyParseResultScenario
    <| sprintf "%s,%s" lResult rResult
    <| sprintf @"\left%sa\right%s" left right

[<Theory>]
[<InlineData(".", ")", true, false)>]
[<InlineData("(", ".", false, true)>]
let ``Empty delimiters should work`` (left : string, right : string, isLeftEmpty : bool, isRightEmpty : bool) =
    verifyParseResultScenario
    <| sprintf "(%s,%s,%A,%A)" left right isLeftEmpty isRightEmpty
    <| sprintf @"\left%sa\right%s" left right

[<Fact>]
let ``Unmatched delimiters should work`` () =
    verifyParseResult @"\left)a\right|"

[<Fact>]
let ``Non-existing delimiter should throw exception`` () =
    let markup = @"\left x\right)"
    assertParseThrows<TexParseException> markup

[<Fact>]
let ``Expression in braces should be parsed`` () =
    verifyParseResult @"\left(2+2\right)"

[<Fact>]
let ``Expression after the braces should be parsed`` () =
    verifyParseResult @"\left(2+2\right) + 1"

[<Fact>]
let ``\text command should be supported`` () =
    verifyParseResult @"\text{test}"

[<Fact>]
let ``Spaces in \text shouldn't be ignored`` () =
    verifyParseResult @"\text{a b c}"

[<Fact>]
let ``\text should support Cyrillic`` () =
    verifyParseResult @"\text{абв}"

[<Fact>]
let ``\mathrm should be parsed properly`` () =
    verifyParseResult @"\mathrm{sin}"

[<Fact>]
let ``\mathrm should be parsed properly for complex eqs`` () =
    verifyParseResult @"\mathrm{\left(2+2\right)} + 1"

[<Fact>]
let ``\mathit should be parsed properly`` () =
    verifyParseResult @"\mathit{sin}"


[<Fact>]
let ``\mathcal should be parsed properly`` () =
    verifyParseResult @"\mathcal{sin}"

[<Fact>]
let ``\mathrm{} should throw exn`` () =
    assertParseThrows<TexParseException> @"\mathrm{}"

[<Fact>]
let ``\lim should be parsed properly`` () =
    verifyParseResult @"\lim_{n} x"

[<Fact>]
let ``{\lim} x should be parsed properly`` () =
    verifyParseResult @"{\lim} x"

[<Fact>]
let ``\sin should be parsed properly`` () =
    verifyParseResult @"\sin^{n} x"

[<Fact>]
let ``\int f should be parsed properly`` () =
    verifyParseResult @"\int f"

[<Fact>]
let ``{} should be parsed properly`` () =
    verifyParseResult @"{}"

[<Fact>]
let ``Delimiter with scripts should be parsed properly`` () =
    verifyParseResult @"\left(2+2\right)_a^b"

let ``\text doesn't create any SymbolAtoms``() =
    verifyParseResult @"\text{2+2}"

[<Fact>]
let ``\sqrt should throw a TexParseException``() =
    assertParseThrows<TexParseException> @"\sqrt"

[<Fact>]
let ``"\sum_ " should throw a TexParseException``() =
    assertParseThrows<TexParseException> @"\sum_ "

[<Theory>]
[<InlineData(@"\color{red}1123");
  InlineData(@"\color{red}{1}123");
  InlineData(@"\color{red} 1123");
  InlineData(@"\color{red} {1}123")>]
let ``\color should parse arguments properly``(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| text

[<Theory>]
[<InlineData(@"\colorbox{red}1123");
  InlineData(@"\colorbox{red}{1}123");
  InlineData(@"\colorbox{red} 1123");
  InlineData(@"\colorbox{red} {1}123")>]
let ``\colorbox should parse arguments properly``(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| text

[<Theory>]
[<InlineData(@"\frac2x123");
  InlineData(@"\frac2{x}123");
  InlineData(@"\frac{2}x123");
  InlineData(@"\frac{2}{x}123");
  InlineData(@"\frac 2 x123");
  InlineData(@"\frac2 {x}123");
  InlineData(@"\frac 2{x}123")>]
let ``\frac should parse arguments properly``(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| text

[<Theory>]
[<InlineData(@"\overline1123");
  InlineData(@"\overline{1}123");
  InlineData(@"\overline 1123");
  InlineData(@"\overline {1}123")>]
let ``\overline should parse arguments properly``(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| text

[<Theory>]
[<InlineData(@"\sqrt1123");
  InlineData(@"\sqrt{1}123");
  InlineData(@"\sqrt 1123");
  InlineData(@"\sqrt {1}123")>]
let ``\sqrt should parse arguments properly``(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| text

[<Theory>]
[<InlineData(@"\sqrt [2]1123");
  InlineData(@"\sqrt [ 2]{1}123");
  InlineData(@"\sqrt[2 ] 1123");
  InlineData(@"\sqrt[ 2 ] {1}123")>]
let ``\sqrt should parse optional argument properly``(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| text

[<Theory>]
[<InlineData(@"\underline1123");
  InlineData(@"\underline{1}123");
  InlineData(@"\underline 1123");
  InlineData(@"\underline {1}123")>]
let ``\underline should parse arguments properly``(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| text

[<Theory>]
[<InlineData("x^y_z");
  InlineData("x^y_{z}");
  InlineData("x^{y}_z");
  InlineData("x^{y}_{z}");
  InlineData("x^y_ z");
  InlineData("x ^ {y} _ {z}")>]
let ``Scripts should be parsed properly``(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| text

[<Theory>]
[<InlineData(@"\text 1123");
  InlineData(@"\text {1}123")>]
let ``\text command should support extended argument parsing``(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| text

[<Fact>]
let ``{\hat T} should parse successfully``() : unit =
    verifyParseResult @"{\hat T}"

[<Fact>]
let ``integral expression should be parsed properly`` () =
    verifyParseResult @"\int_a^b"
