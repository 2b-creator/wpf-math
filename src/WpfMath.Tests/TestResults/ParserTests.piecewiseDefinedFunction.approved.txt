{
  "TextStyle": null,
  "RootAtom": {
    "PreviousAtom": null,
    "Elements": [
      {
        "Character": "f",
        "TextStyle": null,
        "IsDefaultTextStyle": true,
        "IsTextSymbol": false,
        "Type": "Ordinary",
        "Source": {
          "Start": 0,
          "End": 1,
          "Length": 1,
          "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
        }
      },
      {
        "IsDelimeter": true,
        "Name": "(",
        "IsTextSymbol": false,
        "Type": "Opening",
        "Source": {
          "Start": 1,
          "End": 2,
          "Length": 1,
          "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
        }
      },
      {
        "Character": "x",
        "TextStyle": null,
        "IsDefaultTextStyle": true,
        "IsTextSymbol": false,
        "Type": "Ordinary",
        "Source": {
          "Start": 2,
          "End": 3,
          "Length": 1,
          "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
        }
      },
      {
        "IsDelimeter": true,
        "Name": ")",
        "IsTextSymbol": false,
        "Type": "Closing",
        "Source": {
          "Start": 3,
          "End": 4,
          "Length": 1,
          "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
        }
      },
      {
        "IsDelimeter": false,
        "Name": "equals",
        "IsTextSymbol": false,
        "Type": "Relation",
        "Source": {
          "Start": 5,
          "End": 6,
          "Length": 1,
          "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
        }
      },
      {
        "BaseAtom": {
          "MatrixCells": [
            [
              {
                "PreviousAtom": null,
                "Elements": [
                  {
                    "Character": "1",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 14,
                      "End": 15,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "IsDelimeter": false,
                    "Name": "slash",
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 15,
                      "End": 16,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "3",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 16,
                      "End": 17,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  }
                ],
                "Type": "Ordinary",
                "Source": {
                  "Start": 14,
                  "End": 18,
                  "Length": 4,
                  "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                }
              },
              {
                "PreviousAtom": null,
                "Elements": [
                  {
                    "Character": "i",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 20,
                      "End": 21,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "f",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 21,
                      "End": 22,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Type": "Ordinary",
                    "Source": null
                  },
                  {
                    "Character": "0",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 34,
                      "End": 35,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "IsDelimeter": false,
                    "Name": "le",
                    "IsTextSymbol": false,
                    "Type": "Relation",
                    "Source": {
                      "Start": 36,
                      "End": 38,
                      "Length": 2,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "x",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 39,
                      "End": 40,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "IsDelimeter": false,
                    "Name": "le",
                    "IsTextSymbol": false,
                    "Type": "Relation",
                    "Source": {
                      "Start": 41,
                      "End": 43,
                      "Length": 2,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "1",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 44,
                      "End": 45,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "IsDelimeter": false,
                    "Name": "semicolon",
                    "IsTextSymbol": false,
                    "Type": "Punctuation",
                    "Source": {
                      "Start": 45,
                      "End": 46,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  }
                ],
                "Type": "Ordinary",
                "Source": {
                  "Start": 19,
                  "End": 46,
                  "Length": 27,
                  "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                }
              }
            ],
            [
              {
                "PreviousAtom": null,
                "Elements": [
                  {
                    "Character": "2",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 50,
                      "End": 51,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "IsDelimeter": false,
                    "Name": "slash",
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 51,
                      "End": 52,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "3",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 52,
                      "End": 53,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  }
                ],
                "Type": "Ordinary",
                "Source": {
                  "Start": 49,
                  "End": 54,
                  "Length": 5,
                  "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                }
              },
              {
                "PreviousAtom": null,
                "Elements": [
                  {
                    "Character": "i",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 56,
                      "End": 57,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "f",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 57,
                      "End": 58,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Type": "Ordinary",
                    "Source": null
                  },
                  {
                    "Character": "3",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 70,
                      "End": 71,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "IsDelimeter": false,
                    "Name": "le",
                    "IsTextSymbol": false,
                    "Type": "Relation",
                    "Source": {
                      "Start": 72,
                      "End": 74,
                      "Length": 2,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "x",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 75,
                      "End": 76,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "IsDelimeter": false,
                    "Name": "le",
                    "IsTextSymbol": false,
                    "Type": "Relation",
                    "Source": {
                      "Start": 78,
                      "End": 80,
                      "Length": 2,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "4",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 81,
                      "End": 82,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "IsDelimeter": false,
                    "Name": "semicolon",
                    "IsTextSymbol": false,
                    "Type": "Punctuation",
                    "Source": {
                      "Start": 82,
                      "End": 83,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  }
                ],
                "Type": "Ordinary",
                "Source": {
                  "Start": 55,
                  "End": 84,
                  "Length": 29,
                  "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                }
              }
            ],
            [
              {
                "Character": "0",
                "TextStyle": null,
                "IsDefaultTextStyle": true,
                "IsTextSymbol": false,
                "Type": "Ordinary",
                "Source": {
                  "Start": 88,
                  "End": 89,
                  "Length": 1,
                  "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                }
              },
              {
                "PreviousAtom": null,
                "Elements": [
                  {
                    "Character": "e",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 92,
                      "End": 93,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "l",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 93,
                      "End": 94,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "s",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 94,
                      "End": 95,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "e",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 95,
                      "End": 96,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "w",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 96,
                      "End": 97,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "h",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 97,
                      "End": 98,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "e",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 98,
                      "End": 99,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "r",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 99,
                      "End": 100,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "Character": "e",
                    "TextStyle": null,
                    "IsDefaultTextStyle": true,
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 100,
                      "End": 101,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  },
                  {
                    "IsDelimeter": false,
                    "Name": "normaldot",
                    "IsTextSymbol": false,
                    "Type": "Ordinary",
                    "Source": {
                      "Start": 101,
                      "End": 102,
                      "Length": 1,
                      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                    }
                  }
                ],
                "Type": "Ordinary",
                "Source": {
                  "Start": 91,
                  "End": 102,
                  "Length": 11,
                  "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
                }
              }
            ]
          ],
          "VerticalPadding": 0.35,
          "HorizontalPadding": 0.35,
          "MatrixCellAlignment": "Left",
          "Type": "Ordinary",
          "Source": {
            "Start": 14,
            "End": 105,
            "Length": 91,
            "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
          }
        },
        "Type": "Ordinary",
        "Source": {
          "Start": 14,
          "End": 105,
          "Length": 91,
          "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
        }
      }
    ],
    "Type": "Ordinary",
    "Source": {
      "Start": 7,
      "End": 112,
      "Length": 105,
      "Source": "f(x) = \\cases{1/3 & if \\thinspace 0\\le x\\le 1;\\cr 2/3 & if \\thinspace 3\\le x \\le 4; \\cr 0 & elsewhere.\\cr}"
    }
  }
}